crates/ere-openvm/src/lib.rs

```diff
@@ -1,78 +1,133 @@
+    config::{AppConfig, DEFAULT_APP_LOG_BLOWUP, DEFAULT_LEAF_LOG_BLOWUP, SdkVmConfig},
+    keygen::AppProvingKey,
 };
-use openvm_transpiler::elf::Elf;
+use openvm_stark_sdk::config::FriParameters;
+use openvm_transpiler::{elf::Elf, openvm_platform::memory::MEM_SIZE};
+use serde::{Deserialize, Serialize};
+use std::{fs, path::Path, sync::Arc, time::Instant};
 use zkvm_interface::{
     Compiler, Input, InputItem, ProgramExecutionReport, ProgramProvingReport, ProverResourceType,
     zkVM, zkVMError,
 };
 
 include!(concat!(env!("OUT_DIR"), "/name_and_sdk_version.rs"));
 mod error;
-use error::{CompileError, OpenVMError, VerifyError};
 
 #[allow(non_camel_case_types)]
 pub struct OPENVM_TARGET;
 
+#[derive(Clone, Serialize, Deserialize)]
+pub struct OpenVMProgram {
+    elf: Vec<u8>,
+    app_config: AppConfig<SdkVmConfig>,
+}
+
 impl Compiler for OPENVM_TARGET {
     type Error = OpenVMError;
 
-    type Program = Elf;
+    type Program = OpenVMProgram;
 
+    // Inlining `openvm_sdk::Sdk::build` in order to get raw elf bytes.
     fn compile(workspace_path: &Path, guest_relative: &Path) -> Result<Self::Program, Self::Error> {
-        let sdk = Sdk::new();
-
-        // Build the guest crate
-        let elf: Elf = sdk
-            .build(
-                GuestOptions::default(),
-                workspace_path.join(guest_relative),
-                &Default::default(),
-            )
-            .map_err(|e| CompileError::Client(e.into()))?;
-        // TODO: note that this does not transpile (check to see how expensive that is)
+        let guest_directory = workspace_path.join(guest_relative);
+        let pkg = openvm_build::get_package(&guest_directory);
+        let guest_opts = GuestOptions::default().with_profile("release".to_string());
+        let target_dir = match openvm_build::build_guest_package(&pkg, &guest_opts, None, &None) {
+            Ok(target_dir) => target_dir,
+            Err(Some(code)) => return Err(CompileError::BuildFailed(code).into()),
+            Err(None) => return Err(CompileError::BuildSkipped.into()),
+        };
+
+        let elf_path = openvm_build::find_unique_executable(&guest_directory, target_dir, &None)
+            .map_err(|e| CompileError::UniqueElfNotFound(e.into()))?;
+        let elf = fs::read(&elf_path).map_err(|source| CompileError::ReadElfFailed {
+            source,
+            path: elf_path.to_path_buf(),
+        })?;
+
+        let app_config_path = guest_directory.join("openvm.toml");
+        let app_config = if app_config_path.exists() {
+            let toml = fs::read_to_string(&app_config_path).map_err(|source| {
+                CompileError::ReadConfigFailed {
+                    source,
+                    path: app_config_path.to_path_buf(),
+                }
+            })?;
+            toml::from_str(&toml)
+                .map_err(|err| CompileError::DeserializeConfigFailed(err.into()))?
+        } else {
+            AppConfig {
```

Comment:
Do you have any thoughts on why defining this default here and make it part of the program serialization, compared to make it a static var used in `new`? No strong opinions, just curious.

Reply:
I feel the app config is part of compilation (compilation flags thing), and I guess for performance reason most project will have one (to enable certain precompiles), so I guess it'd be okay to store it in `Program`.

If it's a argument in `new`, I'm afraid it'd make host a bit annoying because it needs to check if the guest dir has this config or not.